Automattic/newspack-plugin
# feat(data-events): webhooks

### All Submissions:






As part of the core Data Events system (#2173), webhooks are responsible for integrating the data events to any API that is capable of interpreting the payload.

### Requests

The webhook requests are registered as a custom post type. This allows flexibility to log, query, and more efficiently manage the relationship with registered endpoints.

Once a data event is dispatched, webhook requests may be created if there are endpoints registered to receive that type of data.

Requests make use of WP's post scheduling in order to send a request. Once a request is created, its send is scheduled for 1 minute from now. Every webhook request is expected to have a slight delay but the original timestamp is preserved in the payload.

### Endpoints

A custom taxonomy for registering the webhook URLs and their configuration.

In the next PR, the publisher will be able to register new webhook URLs through the “Connections” wizard. A section at the bottom of the wizard will allow the entire user-facing management:

- List existing endpoints
- Display the latest request response code (endpoint health)
- Enable/disable endpoints
- Add and edit endpoints with their URL, whether it's global or select the data events actions it's supposed to listen to

If we wish to, it’s also possible to display the log of the most recent requests per endpoint.

### Handling request response codes

The registered endpoints are prone to unexpected errors and we should allow the client to fix them while preserving data integrity.

Until the server receives a 2xx response from the client, it attempts using an exponential backoff (2<sup>n</sup> minutes intervals) maxed to 12 hours. On the 15th attempt, an error will kill the request and deactivate the endpoint.

### Cron events

- Twice a day every finished request older than 7 days is permanently deleted
- Every hour ensure to send any request that should've been sent according to its scheduling but it was not

### Tweaks to the data events API

The 3rd argument of `register_listener()` can now be a `callable` **or an `array`**. The array should be a list of strings to be mapped with the hook's arguments, e.g.:

https://github.com/Automattic/newspack-plugin/blob/87b628be7463ddaff2e15d873a779bfdc8ade462/includes/data-events/listeners.php#L79-L86

A listener also no longer dispatches an action if the resulting data is empty.

### Inaugurating listeners

This PR introduces common listeners to be registered at [`includes/data-events/listeners.php`](https://github.com/Automattic/newspack-plugin/blob/feat/data-events-webhooks/includes/data-events/listeners.php). They create the following actions:

 - reader_registered
 - reader_logged_in
 - newsletter_subscribed
 - newsletter_updated

### How to test the changes in this Pull Request:

Confirm all written tests are passing.

1. Visit this site to mock a webhook receiver: https://webhook.site/ and leave this page open
2. Open your site shell with `$ wp shell`:
   1. Copy the endpoint URL and place it in a variable: `$url = 'https://webhook.site/{some-id-here}';`
   2. Create a global endpoint with: `Newspack\Data_Events\Webhooks::create_endpoint( $url, [], true );`
   3. Confirm the endpoint is created by the return of an `int`, which is the endpoint ID
3. Confirm you have RAS active, AMP Plus and newspack log enabled:
```php
define( 'NEWSPACK_AMP_PLUS_ENABLED', true );
define( 'NEWSPACK_EXPERIMENTAL_READER_ACTIVATION', true );
define( 'NEWSPACK_LOG_LEVEL', 2 );
```
4. Visit your site in incognito and go through the reader registration flow
5. Confirm the `NEWSPACK-DATA-EVENTS` and `NEWSPACK-WEBHOOKS` logs:
```
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "reader_logged_in".
[NEWSPACK-WEBHOOKS][Newspack\Data_Events\Webhooks::schedule_request]: Scheduling request 346 for 2022-12-14 13:24:23.
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::handle]: Executing global action handlers for "reader_logged_in".
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::handle]: Executing action handlers for "reader_logged_in".
```
6. Wait one minute until the scheduled request is sent and confirm you than see the following log:
```
[NEWSPACK-WEBHOOKS][Newspack\Data_Events\Webhooks::send_request]: Sending request 346
[NEWSPACK-WEBHOOKS][Newspack\Data_Events\Webhooks::finish_request]: Finishing request 346.
```
7. The https://webhook.site page should be updated with this:
<img width="1017" alt="Captura de Tela 2022-12-14 às 10 26 30" src="https://user-images.githubusercontent.com/820752/207609391-751307a2-48e8-4e26-b532-1d6cc30c71b1.png">

### Other information:
